File: 02_stack/01_min_stack/min_stack.cpp
#include <stack>

using std::stack;

class MinStack
{
  public:
    MinStack()
    {
    }

    void push(int val)
    {
        if (data.empty())
            data.push((Node){.value_ = val, .min_ = val});
        else
        {
            if (val <= data.top().min_)
                data.push((Node){.value_ = val, .min_ = val});
            else
                data.push((Node){.value_ = val, .min_ = data.top().min_});
        }
    }

    void pop()
    {
        data.pop();
    }

    int top()
    {
        if (!data.empty())
            return data.top().value_;
        else
            return 0;
    }

    int getMin()
    {
        if (!data.empty())
            return data.top().min_;
        else
            return 0;
    }

  private:
    struct Node
    {
        int value_;
        int min_;
    };
    stack<Node> data;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

int main()
{
    MinStack obj;

    int val = 3;
    obj.push(val);
    obj.pop();
    int param_3 = obj.top();
    int param_4 = obj.getMin();
}

File: NeetCode/02_stack/01_min_stack/min_stack.py
class MinStack:
    data = []
    def __init__(self):
        self.data.clear()

    def push(self, val: int) -> None:
        if len(self.data) == 0:
            self.data.append((val, val))
        else:
            if val <= self.data[-1][1]:
                self.data.append((val, val))
            else:
                self.data.append((val, self.data[-1][1]))

    def pop(self) -> None:
        self.data.pop()        

    def top(self) -> int:
        if len(self.data) != 0:
            return self.data[-1][0]
        else:
            return 0

    def getMin(self) -> int:
        if len(self.data) != 0:
            return self.data[-1][1]
        else:
            return 0

        


# Your MinStack object will be instantiated and called as such:
val = -1
obj = MinStack()
obj.push(val)
param_3 = obj.top()
param_4 = obj.getMin()
obj.pop()